```
Nama: Dzulfikri Adjmal
Kelompok: 20
```

In [22]:
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.multioutput import MultiOutputClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

from catboost import CatBoostClassifier

import plotly.express as px

df = pd.read_csv('./ai4i2020.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   UDI                      10000 non-null  int64  
 1   Product ID               10000 non-null  object 
 2   Type                     10000 non-null  object 
 3   Air temperature [K]      10000 non-null  float64
 4   Process temperature [K]  10000 non-null  float64
 5   Rotational speed [rpm]   10000 non-null  int64  
 6   Torque [Nm]              10000 non-null  float64
 7   Tool wear [min]          10000 non-null  int64  
 8   Machine failure          10000 non-null  int64  
 9   TWF                      10000 non-null  int64  
 10  HDF                      10000 non-null  int64  
 11  PWF                      10000 non-null  int64  
 12  OSF                      10000 non-null  int64  
 13  RNF                      10000 non-null  int64  
dtypes: float64(3), int64(9)

In [23]:
df.head()

UDI Product ID Type  Air temperature [K]  Process temperature [K]  \
0    1     M14860    M                298.1                    308.6   
1    2     L47181    L                298.2                    308.7   
2    3     L47182    L                298.1                    308.5   
3    4     L47183    L                298.2                    308.6   
4    5     L47184    L                298.2                    308.7   

   Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  Machine failure  TWF  \
0                    1551         42.8                0                0    0   
1                    1408         46.3                3                0    0   
2                    1498         49.4                5                0    0   
3                    1433         39.5                7                0    0   
4                    1408         40.0                9                0    0   

   HDF  PWF  OSF  RNF  
0    0    0    0    0  
1    0    0    0    0  
2    0    0    0    0  
3    0    0    0    0  
4    0    0    0    0

In [24]:
df.describe()

UDI  Air temperature [K]  Process temperature [K]  \
count  10000.00000         10000.000000             10000.000000   
mean    5000.50000           300.004930               310.005560   
std     2886.89568             2.000259                 1.483734   
min        1.00000           295.300000               305.700000   
25%     2500.75000           298.300000               308.800000   
50%     5000.50000           300.100000               310.100000   
75%     7500.25000           301.500000               311.100000   
max    10000.00000           304.500000               313.800000   

       Rotational speed [rpm]   Torque [Nm]  Tool wear [min]  Machine failure  \
count            10000.000000  10000.000000     10000.000000     10000.000000   
mean              1538.776100     39.986910       107.951000         0.033900   
std                179.284096      9.968934        63.654147         0.180981   
min               1168.000000      3.800000         0.000000         0.000000   
25%               1423.000000     33.200000        53.000000         0.000000   
50%               1503.000000     40.100000       108.000000         0.000000   
75%               1612.000000     46.800000       162.000000         0.000000   
max               2886.000000     76.600000       253.000000         1.000000   

                TWF           HDF           PWF           OSF          RNF  
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.00000  
mean       0.004600      0.011500      0.009500      0.009800      0.00190  
std        0.067671      0.106625      0.097009      0.098514      0.04355  
min        0.000000      0.000000      0.000000      0.000000      0.00000  
25%        0.000000      0.000000      0.000000      0.000000      0.00000  
50%        0.000000      0.000000      0.000000      0.000000      0.00000  
75%        0.000000      0.000000      0.000000      0.000000      0.00000  
max        1.000000      1.000000      1.000000      1.000000      1.00000

In [25]:
df.nunique()

UDI                        10000
Product ID                 10000
Type                           3
Air temperature [K]           93
Process temperature [K]       82
Rotational speed [rpm]       941
Torque [Nm]                  577
Tool wear [min]              246
Machine failure                2
TWF                            2
HDF                            2
PWF                            2
OSF                            2
RNF                            2
dtype: int64

In [26]:
df['Type'].value_counts()

Type
L    6000
M    2997
H    1003
Name: count, dtype: int64

In [27]:
target_column= df.iloc[:, 8:].columns.to_list()

for i in target_column:
    print(df[i].value_counts())

Machine failure
0    9661
1     339
Name: count, dtype: int64
TWF
0    9954
1      46
Name: count, dtype: int64
HDF
0    9885
1     115
Name: count, dtype: int64
PWF
0    9905
1      95
Name: count, dtype: int64
OSF
0    9902
1      98
Name: count, dtype: int64
RNF
0    9981
1      19
Name: count, dtype: int64


In [28]:
encoding = LabelEncoder()

df['Type'] = encoding.fit_transform(df['Type'])
df['Type'].value_counts()

Type
1    6000
2    2997
0    1003
Name: count, dtype: int64

## Penyelesaian 1

In [29]:
X = df.drop(['UDI', 'Product ID', 'Machine failure', 'TWF', 'HDF', 'PWF', 'OSF', 'RNF'], axis=1)
y = df.iloc[:, 8:]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) # Membagi data dengan rasio 80:20, 80% data training dan 20% data testing
print("X train shape: ", X_train.shape)
print("X test shape: ", X_test.shape)
print("y train shape: ", y_train.shape)
print("y test shape: ", y_test.shape)

X train shape:  (8000, 6)
X test shape:  (2000, 6)
y train shape:  (8000, 6)
y test shape:  (2000, 6)


In [30]:
model_LR = MultiOutputClassifier(LogisticRegression())
model_LR.fit(X_train, y_train)

y_pred = model_LR.predict(X_test)

print("Accuracy: ", accuracy_score(y_test, y_pred))


Accuracy:  0.9655


In [31]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.62      0.26      0.37        61
           1       0.00      0.00      0.00        11
           2       0.75      0.35      0.48        17
           3       0.94      0.80      0.86        20
           4       0.74      0.78      0.76        18
           5       0.00      0.00      0.00         6

   micro avg       0.73      0.39      0.51       133
   macro avg       0.51      0.37      0.41       133
weighted avg       0.62      0.39      0.46       133
 samples avg       0.02      0.01      0.01       133



c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.

c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning:

Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.

c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning:

F-score is ill-defined and being set

## Penyelesaian 2

### Praproses Data

In [ ]:
df.isna().sum()

UDI                        0
Product ID                 0
Type                       0
Air temperature [K]        0
Process temperature [K]    0
Rotational speed [rpm]     0
Torque [Nm]                0
Tool wear [min]            0
Machine failure            0
TWF                        0
HDF                        0
PWF                        0
OSF                        0
RNF                        0
dtype: int64

### Memeriksa data duplikat

In [ ]:
df.duplicated().sum()

0

In [ ]:
data_w_failure = df[df['Machine failure'] == 1]
data_w_failure

UDI Product ID  Type  Air temperature [K]  Process temperature [K]  \
50      51     L47230     1                298.9                    309.1   
69      70     L47249     1                298.9                    309.0   
77      78     L47257     1                298.8                    308.9   
160    161     L47340     1                298.4                    308.2   
161    162     L47341     1                298.3                    308.1   
...    ...        ...   ...                  ...                      ...   
9758  9759     L56938     1                298.6                    309.8   
9764  9765     L56944     1                298.5                    309.5   
9822  9823     L57002     1                298.5                    309.4   
9830  9831     L57010     1                298.3                    309.3   
9974  9975     L57154     1                298.6                    308.2   

      Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  Machine failure  \
50                      2861          4.6              143                1   
69                      1410         65.7              191                1   
77                      1455         41.3              208                1   
160                     1282         60.7              216                1   
161                     1412         52.3              218                1   
...                      ...          ...              ...              ...   
9758                    2271         16.2              218                1   
9764                    1294         66.7               12                1   
9822                    1360         60.9              187                1   
9830                    1337         56.1              206                1   
9974                    1361         68.2              172                1   

      TWF  HDF  PWF  OSF  RNF  
50      0    0    1    0    0  
69      0    0    1    1    0  
77      1    0    0    0    0  
160     0    0    0    1    0  
161     0    0    0    1    0  
...   ...  ...  ...  ...  ...  
9758    1    0    0    0    0  
9764    0    0    1    0    0  
9822    0    0    0    1    0  
9830    0    0    0    1    0  
9974    0    0    1    1    0  

[339 rows x 14 columns]

### Melakukan prediksi menggunakan model Random Forest

In [ ]:
model_RF = RandomForestClassifier(random_state=42)
multi_target = MultiOutputClassifier(model_RF, n_jobs=-1)

multi_target.fit(X_train, y_train)

y_pred = multi_target.predict(X_test)


In [ ]:
print("Accuracy: ", accuracy_score(y_test, y_pred))

Accuracy:  0.978


In [ ]:
print(classification_report(y_test, y_pred, target_names=target_column))

                 precision    recall  f1-score   support

Machine failure       0.84      0.59      0.69        61
            TWF       0.00      0.00      0.00        11
            HDF       1.00      0.76      0.87        17
            PWF       0.94      0.75      0.83        20
            OSF       0.91      0.56      0.69        18
            RNF       0.00      0.00      0.00         6

      micro avg       0.89      0.56      0.69       133
      macro avg       0.61      0.44      0.51       133
   weighted avg       0.78      0.56      0.65       133
    samples avg       0.02      0.02      0.02       133



c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{

## Penyelesaian 3 (Catboost)

In [ ]:
model_cb = CatBoostClassifier(iterations=1000, learning_rate=0.1, verbose=0, random_state=42, allow_writing_files=False)
multi_target_cb = MultiOutputClassifier(model_cb, n_jobs=-1)

multi_target_cb.fit(X_train, y_train)

y_pred = multi_target_cb.predict(X_test)

In [ ]:
print('Accuracy: ', accuracy_score(y_test, y_pred))

Accuracy:  0.9785


In [ ]:
print(classification_report(y_test, y_pred, target_names=target_column))

                 precision    recall  f1-score   support

Machine failure       0.86      0.62      0.72        61
            TWF       0.00      0.00      0.00        11
            HDF       1.00      0.76      0.87        17
            PWF       0.79      0.75      0.77        20
            OSF       0.88      0.83      0.86        18
            RNF       0.00      0.00      0.00         6

      micro avg       0.87      0.61      0.72       133
      macro avg       0.59      0.50      0.54       133
   weighted avg       0.76      0.61      0.67       133
    samples avg       0.02      0.02      0.02       133



c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Asus\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{

## Kesimpulan

Ketiga model yang dibuat memiliki akurasi yang tidak jauh berbeda. Model Logistic Regression memiliki akurasi 96%, model Random Forest memiliki akurasi 97%, dan model Catboost memiliki akurasi 97%. Namun terdapat perbedaan yang cukup signifikan jika dilihat dari nilai precision dari masing-masing model.

| Model | Label | Precision|
| --- | --- | --- |
| Logistic Regression | 0 | 0.62 |
| Logistic Regression | 1 | 0.00 |
| Logistic Regression | 2 | 0.75 |
| Logistic Regression | 3 | 0.94 |
| Logistic Regression | 4 | 0.74 |
| Logistic Regression | 5 | 0.00 |
| --- | --- | --- |
| Random Forest | 0 | 0.84 |
| Random Forest | 1 | 0.00 |
| Random Forest | 2 | 1.00 |
| Random Forest | 3 | 0.94 |
| Random Forest | 4 | 0.91 |
| Random Forest | 5 | 0.00 |
| --- | --- | --- |
| Catboost | 0 | 0.86 |
| Catboost | 1 | 0.00 |
| Catboost | 2 | 1.00 |
| Catboost | 3 | 0.79 |
| Catboost | 4 | 0.88 |
| Catboost | 5 | 0.00 |

Note: 0 = Machine failure, 1 = TWF, 2 = HDF, 3 = PWF, 4 = OSF, 5 = RNF.

Model Random Forest memiliki nilai precision yang lebih baik dibandingkan dengan kedua model lainnya. Model Random Forest memiliki nilai precision yang lebih baik pada label 2, 3, dan 5.